File: 001.two-sum/Solution.java
/*
https://leetcode.com/problems/two-sum/

Given an array of integers, return indices of the two numbers such that they
add up to a specific target.

You may assume that each input would have exactly one solution, and you may
not use the same element twice.
*/
import java.util.*;

public class Solution {

    public int[] twoSum(int[] nums, int target) {
        return first(nums, target);
    }

    private int[] first(int[] nums, int target) {
        int[] output = new int[2];
        Map<Integer, Integer> seen = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            if (seen.containsKey(target-nums[i])) {
                output[0] = seen.get(target-nums[i]);
                output[1] = i;
                break;
            }
            seen.put(nums[i], i);
        }

        return output;
    }

    public static void main(String[] args) {

    }
}

File: 001.two-sum/solution.js
/**
 * Finds if there exists two numbers in the input array that sum to target.
 * @param {*} nums - Array of integers.
 * @param {*} target - Target sum.
 * @return {number[]} - Indices of numbers which sum to target.
 */
function twoSum(nums, target) {
  let map = new Map();

  for (let i = 0; i < nums.length; i++) {
    const curr = nums[i];
    const temp = target - curr;

    if (map.has(temp)) {
      return [map.get(temp), i];
    }

    map.set(curr, i);
  }
}

(function run() {
  let tests = [
    { nums: [2, 7, 11, 15], target: 9, expected: [0, 1] },
    { nums: [3, 2, 4], target: 6, expected: [1, 2] },
    { nums: [3, 3], target: 6, expected: [0, 1] }
  ];

  tests.forEach(test => {
    console.log('Running test...');
    const result = twoSum(test.nums, test.target);
    const passed = result.every((value, index) => {
      return value === test.expected[index];
    });
    console.log(`Test passed? : ${passed}`);
  });
})();